common/src/autogluon/common/utils/cpu_utils.py

```diff
@@ -0,0 +1,175 @@
+                    logger.debug(f"Detected cgroup v1 CPU limit: {cgroup_count}")
+                    return cgroup_count
+        except (IOError, ValueError) as e:
+            logger.debug(f"Error reading cgroup v1 CPU limit: {e}")
+
+    return os_cpu_count
+
+
+def get_available_cpu_count(only_physical_cores=False):
+    """
+    Get the number of available CPU cores, respecting container limits,
+    CPU affinity, and environment variables.
+
+    This function checks multiple sources to determine the true number of CPUs
+    available to the current process:
+
+    1. Environment variables (AG_CPU_COUNT, SLURM_CPUS_PER_TASK)
+    2. Loky's CPU detection
+    3. CPU affinity via os.sched_getaffinity
+    4. CPU affinity via psutil
+    5. cgroup limits (for Docker containers)
+    6. Default multiprocessing.cpu_count()
+
+    Parameters
+    ----------
+    only_physical_cores : bool, default=False
+        If True, detects only physical CPU cores (not including hyperthreading/SMT).
+        This can be beneficial for CPU-intensive tasks like time series forecasting
+        where physical cores often provide better performance than logical cores.
+
+    Returns
+    -------
+    int
+        The number of available CPU cores. Returns the minimum count from
+        all successful detection methods to avoid CPU oversubscription.
+    """
+    # Start with system reported CPU count as the default
+    if only_physical_cores:
+        try:
+            default_cpu_count = psutil.cpu_count(logical=False)
+            if default_cpu_count is None:
+                # Fallback to logical cores if physical count unavailable
+                default_cpu_count = multiprocessing.cpu_count()
+        except Exception:
+            default_cpu_count = multiprocessing.cpu_count()
+
+    else:
+        default_cpu_count = multiprocessing.cpu_count()
+
+    available_counts = [default_cpu_count]
+
+    # Log all detected values for debugging
+    core_type = "physical" if only_physical_cores else "logical"
+    logger.debug(f"System default {core_type} CPU count: {default_cpu_count}")
+
+    # 1. Check environment variables first (highest priority)
+    env_var_names = ["AG_CPU_COUNT", "SLURM_CPUS_PER_TASK"]
+    for var_name in env_var_names:
+        if var_name in os.environ:
+            try:
+                count = int(os.environ[var_name])
+                if count > 0:
+                    logger.debug(f"{var_name} environment variable detected: {count}")
+                    # Environment variable overrides everything
+                    return count
+            except ValueError:
+                pass
+
+    # 2. Try loky's CPU count which handles various container scenarios
+    loky_count = loky.cpu_count()
+    logger.debug(f"loky.cpu_count(): {loky_count}")
+    # Note: loky doesn't distinguish between physical/logical cores, so we use it as-is
+    # for logical cores, but skip it for physical cores to avoid mixing different types
```

Comment:
I'm a bit confused about this comment. Why cannot we call [`loky.cpu_count(only_physical_cores=only_physical_cores)`](https://github.com/joblib/loky/blob/15dfcd77faadfb853841162696b486233e540b0b/loky/backend/context.py#L78) if `only_physical_cores=True` in our method?

Reply:
Good call. I am not aware that loky takes physical_core as parameter as well